src/main/java/bio/terra/service/dataset/flight/ingest/IngestParseJsonFileStep.java

```diff
@@ -0,0 +1,103 @@
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class IngestParseJsonFileStep implements Step {
+
+  private final GcsPdao gcsPdao;
+  private final Dataset dataset;
+  private final ObjectMapper objectMapper;
+
+  public IngestParseJsonFileStep(GcsPdao gcsPdao, ObjectMapper objectMapper, Dataset dataset) {
+    this.gcsPdao = gcsPdao;
+    this.objectMapper = objectMapper;
+    this.dataset = dataset;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    IngestRequestModel ingestRequest = IngestUtils.getIngestRequestModel(flightContext);
+    List<String> gcsFileLines =
+        gcsPdao.getGcsFilesLines(
+            ingestRequest.getPath(), dataset.getProjectResource().getGoogleProjectId());
+    List<String> fileRefColumnNames =
+        dataset.getTableByName(ingestRequest.getTable()).orElseThrow().getColumns().stream()
+            .filter(c -> c.getType() == TableDataType.FILEREF)
+            .map(Column::getName)
+            .collect(Collectors.toList());
+    var workingMap = flightContext.getWorkingMap();
+    workingMap.put(IngestMapKeys.TABLE_SCHEMA_FILE_COLUMNS, fileRefColumnNames);
+
+    List<String> errors = new ArrayList<>();
+    List<JsonNode> fileLineJson =
+        gcsFileLines.stream()
+            .map(
+                content -> {
+                  try {
+                    return objectMapper.readTree(content);
+                  } catch (JsonProcessingException ex) {
+                    errors.add(ex.getMessage());
+                    return null;
+                  }
+                })
+            .collect(Collectors.toList());
+
+    if (!errors.isEmpty()) {
+      IngestFailureException ex =
+          new IngestFailureException(
+              "Ingest control file at " + ingestRequest.getPath() + " could not be processed",
+              errors);
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_FATAL, ex);
+    }
+
+    workingMap.put(IngestMapKeys.BULK_LOAD_JSON_LINES, fileLineJson);
+    Set<BulkLoadFileModel> bulkLoadFileModels =
+        fileLineJson.stream()
+            .flatMap(
+                node ->
+                    fileRefColumnNames.stream()
+                        .map(
+                            columnName -> {
+                              JsonNode fileRefNode = node.get(columnName);
+                              if (fileRefNode.isObject()) {
+                                return Optional.of(
+                                    objectMapper.convertValue(
+                                        fileRefNode, BulkLoadFileModel.class));
+                              } else {
+                                return Optional.<BulkLoadFileModel>empty();
+                              }
+                            })
+                        .filter(Optional::isPresent)
+                        .map(Optional::get))
```

Comment:
`Optional` doesn't seem to gain much here IMO. Using `null` directly is less code:
```suggestion
                    fileRefColumnNames.stream()
                        .map(
                            columnName -> {
                              JsonNode fileRefNode = node.get(columnName);
                              if (fileRefNode.isObject()) {
                                return 
                                    objectMapper.convertValue(
                                        fileRefNode, BulkLoadFileModel.class);
                              }
                              return null;
                            })
                        .filter(Objects::nonNull))
```

Reply:
Very true. I'm still in Scala state of mind I guess.